tools/githelper.py

```diff
@@ -0,0 +1,111 @@
+import itertools
+import re
+import logging
+import subprocess
+import typing
+
+
+Line = typing.NamedTuple(
+    "Line",
+    [("sourcefile", str), ("number", int), ("text", str), ("added", bool)],
+)
+FileLines = typing.NamedTuple(
+    "FileLines",
+    [("filename", str), ("lines", typing.Sequence[typing.Tuple[int, int]])],
+)
+
+
+def get_toplevel_path() -> str:
+    logger = logging.getLogger(__name__)
+
+    cmd = ["git", "rev-parse", "--show-toplevel"]
+    logger.debug("Executing: %r", cmd)
+    return subprocess.check_output(cmd, text=True).strip()
+
+
+def get_changed_lines(
+    from_ref=None, filter_lines=None, include_files=None
+) -> typing.Iterable[Line]:
+    """Inspect `git diff` output, yields changed lines."""
+
+    logger = logging.getLogger(__name__)
+
+    cmd = ["git", "diff", "--unified=0", from_ref if from_ref else "HEAD"]
```

Comment:
We cannot only use git diff, because in case of merge commits, the automatic merged lines are also reformatted.
In merge commits, we only want to check the resolution of merge conflicts. 
This can be done with "git show" 

Lines with conflict resolutions are marked with ++ or --

I think we can only do it with a single revision.

Reply:
I think this was original done with the  
"git status --porcelain" call

Reply:
I don't think this is worth the hassle, for various reasons:
1. This is [what the original git-clang-format does, too](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format#L277) and wasn't aware of any issue with that.
2. `git show` will only work for a single commit, so implementing `--from-ref`/`PRE_COMMIT_FROM_REF` will be a mess (and we adding support for that is the whole point of our custom script)
3. For FF-only merges, this won't be an issue
4. For merges this would re-run clang-format on the changed lines, but tbh all those lines should be pre-commit-clean anyway.

Reply:
1. We already had strong issues with it in the PR of @hacksdump. 
Due to a merge with master many unrelated files have been reformatted. This must not happen, because it makes a PR hard to review.

2. We have no use case for this, so it is fine to use it fro single commits only. At the users machine the user will commit only a single merge commit, so it is fine. In CI check it will check the PR against the master, so intermediate master merges are not considered. 

3. right 

4. No, If we merge master we merge legacy lines which are not clean. We must only check lines with solved conflicts.

Reply:
> We have no use case for this, so it is fine to use it fro single commits only. At the users machine the user will commit only a single merge commit, so it is fine. In CI check it will check the PR against the master, so intermediate master merges are not considered.

Using `git show` is not possible. You can only use it to show a commit's diff, but this runs *before* committing, so we need to compare the working tree to a commit. I looked into the git documentation, but I see no real alternative.

> No, If we merge master we merge legacy lines which are not clean. We must only check lines with solved conflicts.

This is only true for very old branches that didn't merge master since we introduced the clang-format pre-commit hook in January. Any code committed since then needs to be pre-commit clean anyway. I'm not sure that we even have such branches that are likely to be merged, and if so you can just `SKIP=clang-format git pull upstream master` for the merge.

> We already had strong issues with it in the PR of @hacksdump.
Due to a merge with master many unrelated files have been reformatted. This must not happen, because it makes a PR hard to review.

In this case it's probably because the branch was old, so the merge included commits that were not formatted properly. I think this is unfortunate, but not critical and the `clang-format` hook can be skipped in that case.